### PR TITLE
[1LP][RFR] Add OpenstackNodeCollection to fix host lifecycle tests

### PR DESCRIPTION
**Pull request by jadhaj**
Purpose or Intent
=================
Add OpenstackNodeCollection on openstack_node.py
Update entry on setup.py
Modify host_lifecycle tests fixture to use new collection

{{ pytest: cfme/tests/openstack/infrastructure/test_host_lifecycle.py }}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit e2bf0fa9730b068c84859bed658083f1570d343d:

`cfme/infrastructure/openstack_node.py`:
- :bangbang: Line 41:1: **E302** *expected 2 blank lines, found 1*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/infrastructure/openstack_node.py [hunk 1]
```diff
@@ -2,7 +2,13 @@
 """A model of an Openstack Infrastructure Node in CFME."""
 
 from cfme.infrastructure.host import Host
-from cfme.utils.appliance.implementations.ui import navigate_to
+import attr
+from cfme.common.host_views import HostsView
+from cfme.utils.appliance.implementations.ui import CFMENavigateStep, navigate_to, navigator
+from navmazing import NavigateToAttribute
+from selenium.common.exceptions import NoSuchElementException
```
Comment (mshriver) [inline cfme/infrastructure/openstack_node.py]: **Required**
Pull this from widgetastic.exceptions instead please.

File: cfme/infrastructure/openstack_node.py [hunk 2]
```diff
@@ -2,7 +2,13 @@
 """A model of an Openstack Infrastructure Node in CFME."""
 
 from cfme.infrastructure.host import Host
-from cfme.utils.appliance.implementations.ui import navigate_to
+import attr
```
Comment (mshriver) [inline cfme/infrastructure/openstack_node.py]: **Required**

Import statements need to be organized and alphabetized:
http://cfme-tests.readthedocs.io/guides/dev_guide.html#code-style

File: cfme/infrastructure/openstack_node.py [hunk 3]
```diff
@@ -32,3 +38,31 @@ def run_introspection(self):
...
+
+
+@attr.s
+class OpenstackNodeCollection(HostCollection):
+    """Collection object for the :py:class:`cfme.infrastructure.openstack_node.OpenstackNode`."""
+
+    ENTITY = OpenstackNode
+
+
+@navigator.register(OpenstackNodeCollection)
```
Comment (mshriver) [inline cfme/infrastructure/openstack_node.py]: This is duplicating the 'All' destination registered against HostCollection, and shouldn't be necessary.

https://github.com/ManageIQ/integration_tests/blob/master/cfme/infrastructure/host.py#L571

File: cfme/infrastructure/openstack_node.py [hunk 4]
```diff
@@ -32,3 +38,31 @@ def run_introspection(self):
...
+    def step(self):
+        try:
+            self.prerequisite_view.navigation.select("Compute", "Infrastructure", "Hosts")
+        except NoSuchElementException:
+            self.prerequisite_view.navigation.select("Compute", "Infrastructure", "Nodes")
+
+
+@navigator.register(OpenstackNode)
+class Details(CFMENavigateStep):
+    VIEW = HostDetailsView
```
Comment (mshriver) [inline cfme/infrastructure/openstack_node.py]: Again, navigation is duplicate of 'Details' destination registered against Host, shouldn't be necessary.

https://github.com/ManageIQ/integration_tests/blob/master/cfme/infrastructure/host.py#L583

File: cfme/infrastructure/openstack_node.py [hunk 5]
```diff
@@ -2,7 +2,13 @@
 """A model of an Openstack Infrastructure Node in CFME."""
 
 from cfme.infrastructure.host import Host
-from cfme.utils.appliance.implementations.ui import navigate_to
+import attr
+from cfme.common.host_views import HostsView
+from cfme.utils.appliance.implementations.ui import CFMENavigateStep, navigate_to, navigator
+from navmazing import NavigateToAttribute
+from selenium.common.exceptions import NoSuchElementException
+from cfme.common.host_views import HostDetailsView
```
Comment (mshriver) [inline cfme/infrastructure/openstack_node.py]: Combine with the HostsView import in one statement.

File: cfme/tests/openstack/infrastructure/test_host_lifecycle.py [hunk 6]
```diff
@@ -16,12 +16,14 @@
 @pytest.fixture(scope='module')
 def host(provider):
     """Find a host for test scenario"""
-    view = navigate_to(OpenstackNode, 'All')
-    hosts = view.entities.get_all()
+    host_collection = provider.appliance.collections.openstack_nodes
+    hosts = host_collection.all(provider)
```
Comment (mshriver) [inline cfme/tests/openstack/infrastructure/test_host_lifecycle.py]: This is where `.filter()` _would_ be used if `all()` was implemented correctly (just FYI, nothing for you to fix here unless you're feeling adventurous.

**Comment by dajoRH** [comment]
Lint report for commit bbfe310bb806a0b50bc1ed2bcd9a65fbc1ac5bfe:

`cfme/infrastructure/openstack_node.py`:
- :red_circle: Line 5:1: **F401** *'cfme.common.host_views.HostDetailsView' imported but unused*
- :red_circle: Line 5:1: **F401** *'cfme.common.host_views.HostsView' imported but unused*
- :red_circle: Line 7:1: **F401** *'cfme.utils.appliance.implementations.ui.CFMENavigateStep' imported but unused*
- :red_circle: Line 7:1: **F401** *'cfme.utils.appliance.implementations.ui.navigator' imported but unused*
- :red_circle: Line 8:1: **F401** *'navmazing.NavigateToAttribute' imported but unused*
- :red_circle: Line 9:1: **F401** *'widgetastic.exceptions.NoSuchElementException' imported but unused*
- :heavy_exclamation_mark: Line 44:27: **W292** *no newline at end of file*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 72e3b275bff3e0ebacb56b91c15627e64cd53644:

`cfme/infrastructure/openstack_node.py`:
- :heavy_exclamation_mark: Line 42:27: **W292** *no newline at end of file*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ade844a99631f684570212d6157ca6cb87a9c9a4:

`cfme/infrastructure/openstack_node.py`:
- :heavy_exclamation_mark: Line 43:1: **W293** *blank line contains whitespace*
- :heavy_exclamation_mark: Line 43:1: **W391** *blank line at end of file*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/openstack/infrastructure/test_host_lifecycle.py [hunk 7]
```diff
@@ -74,7 +77,7 @@ def test_register_host(provider, host, has_mistral_service):
     Metadata:
         test_flag: openstack_scale"""
     hosts_before = [h.uuid for h in provider.mgmt.iapi.node.list()]
-    provider.register(provider.data['instackenv_file_path'])
+    provider.register('instackenv.json')
```
Comment (mshriver) [inline cfme/tests/openstack/infrastructure/test_host_lifecycle.py]: Add comment describing where this string is coming from

File: cfme/infrastructure/openstack_node.py [hunk 8]
```diff
@@ -1,7 +1,8 @@
 # -*- coding: utf-8 -*-
 """A model of an Openstack Infrastructure Node in CFME."""
 
-from cfme.infrastructure.host import Host
+import attr
+from cfme.infrastructure.host import Host, HostCollection
```
Comment (mshriver) [inline cfme/infrastructure/openstack_node.py]: empty line between pip installed and local imports please.

```python
import attr

from cfme.infrastructure.host import stuff
```

**Comment by dajoRH** [comment]
Lint report for commit cf60ed7b6251f490875149491121a7623d808e0e:

`cfme/tests/openstack/infrastructure/test_host_lifecycle.py`:
- :heavy_exclamation_mark: Line 82:72: **W291** *trailing whitespace*
- :heavy_exclamation_mark: Line 83:76: **W291** *trailing whitespace*
- :heavy_exclamation_mark: Line 84:78: **W291** *trailing whitespace*
- :heavy_exclamation_mark: Line 85:82: **W291** *trailing whitespace*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/openstack/infrastructure/test_host_lifecycle.py (hunk 6 above)
Comment (psav) [reply cfme/tests/openstack/infrastructure/test_host_lifecycle.py]: **REQUIRED:** Please add a TODO about this here

File: cfme/tests/openstack/infrastructure/test_host_lifecycle.py [hunk 9]
```diff
@@ -74,7 +77,14 @@ def test_register_host(provider, host, has_mistral_service):
     Metadata:
         test_flag: openstack_scale"""
     hosts_before = [h.uuid for h in provider.mgmt.iapi.node.list()]
-    provider.register(provider.data['instackenv_file_path'])
+
+    """instackenv.json is a file containing the node(s) to be registered.
```
Comment (mshriver) [inline cfme/tests/openstack/infrastructure/test_host_lifecycle.py]: **Required**

I disagree with this change, now that I understand exactly what you're designing here. This should absolutely still be pulled from the provider yaml data, and not hard coded within a test function.

Comment (mshriver) [reply cfme/tests/openstack/infrastructure/test_host_lifecycle.py]: If anything, a `python.path.local` object should be created in order to verify the file exists before proceeding, and skipping the test if it does not.

**Pull request merged**